index.d.ts

```diff
@@ -61,10 +75,72 @@ type EncodingOption =
 type DefaultEncodingOption = 'utf8';
 type BufferEncodingOption = 'buffer';
 
-type GetStdoutStderrType<EncodingType extends EncodingOption> =
-  EncodingType extends DefaultEncodingOption ? string : Uint8Array;
+type TupleItem<
+	Tuple extends readonly unknown[],
+	Index extends string,
+> = Tuple[Index extends keyof Tuple ? Index : number];
+
+// Whether `stdout|stderr: 'ignore', 'inherit'`, etc. is used
+type IgnoresStreamResult<
+	StreamIndex extends string,
+	OptionsType extends Options = Options,
+> = StreamIndex extends '0' ? false
+	: OptionsType[TupleItem<StdioOptionNames, StreamIndex>] extends NoOutputStdioOption ? true
+		: OptionsType['stdio'] extends NoOutputStdioOption ? true
+			: OptionsType['stdio'] extends StdioOptionsArray
+				? (TupleItem<OptionsType['stdio'], StreamIndex> extends NoOutputStdioOption ? true : false)
+				: false;
+
+type StdioOptionNames = ['stdin', 'stdout', 'stderr'];
+
+// Whether `result.stdout|stderr|all` is `undefined`
+type IgnoresStreamOutput<
+	StreamIndex extends string,
+	OptionsType extends Options = Options,
+> = OptionsType extends {buffer: false} ? true
+	: IgnoresStreamResult<StreamIndex, OptionsType> extends true ? true : false;
+
+// Type of `result.stdout|stderr`
+type StdioOutput<
+	StreamIndex extends string,
+	OptionsType extends Options = Options,
+> = IgnoresStreamOutput<StreamIndex, OptionsType> extends true
+	? undefined
+	: StreamResult<OptionsType>;
+
+type StreamResult<OptionsType extends Options = Options> =
+	// Default value for `encoding`, when `OptionsType` is `{}`
+	unknown extends OptionsType['encoding'] ? string
+		// Any value for `encoding`, when `OptionsType` is `Options`
+		: EncodingOption extends OptionsType['encoding'] ? Uint8Array | string
+			// `encoding: buffer`
+			: OptionsType['encoding'] extends 'buffer' ? Uint8Array
+				// `encoding: not buffer`
+				: string;
+
+// Type of `result.all`
+type AllOutput<OptionsType extends Options = Options> = IgnoresStreamOutput<'1', OptionsType> extends true
+	? StdioOutput<'2', OptionsType>
+	: StdioOutput<'1', OptionsType>;
+
+// Type of `result.stdio`
+type StdioArrayOutput<OptionsType extends Options = Options> = MapStdioOptions<
+OptionsType['stdio'] extends StdioOptionsArray ? OptionsType['stdio'] : ['pipe', 'pipe', 'pipe'],
+OptionsType
+>;
+
+type MapStdioOptions<
+	StdioOptionsArrayType extends StdioOptionsArray,
+	OptionsType extends Options = Options,
+> = {
+	[StreamIndex in keyof StdioOptionsArrayType & string]: StreamIndex extends '0'
+		? undefined
+		: StreamIndex extends '1' | '2'
+			? StdioOutput<StreamIndex, OptionsType>
+			: StdioOutput<StreamIndex, OptionsType> | undefined
```

Comment:
Unlike `result.stdio[1]` and `result.stdio[2]`, `result.stdio[3]` (and later ones) can be possible `undefined` when the file descriptor is used for input instead of output. In other words, when it behaves like `stdin` not like `stdout`/`stderr`. For example, this happens when passing an `Uint8Array` as input.